### PR TITLE
Added contact resistance to model.

### LINKED ISSUE #4140: Add contact resistance heat source
Add a heat source due to the parameter "Contact resistance [Ohm]"

### PR DESCRIPTION
**Pull request by parkec3**
# Description

Added a term for the contact resistance within heating. Still needs to be tested for functionality as well as the typical tests. The contact resistance parameter within pybamm is [ohm] instead of [ohm m2], so I left off the surface area when calculating the heating term. The volume parameter also needs to be checked to ensure it's the correct volume calculation.

Fixes #4140

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/models/submodels/thermal/base_thermal.py [hunk 1]
```diff
@@ -175,15 +175,25 @@ def _get_standard_coupled_variables(self, variables):
             Q_rev_n, pybamm.FullBroadcast(0, "separator", "current collector"), Q_rev_p
         )
 
+        # Contact resistance heating Q_cr
+        # TODO check that the volume is the correct volume for contact resistance
+        if self.options["contact resistance"] == "true":
+                I = variables["Current [A]"]
+                V = param.V_cell
+                Q_cr = I**2 * param.R_contact / V
+            else:
```
Comment (valentinsulzer) [inline pybamm/models/submodels/thermal/base_thermal.py:184]: `else` is not aligned with the `if` statement so the tests won't run

Comment (valentinsulzer) [reply pybamm/models/submodels/thermal/base_thermal.py:184]: you can install the "ruff" linter in vscode to highlight things like this

File: pybamm/models/submodels/thermal/base_thermal.py [hunk 2]
```diff
@@ -216,6 +229,7 @@ def _get_standard_coupled_variables(self, variables):
         Q_ohm_vol_av = Q_ohm_W / V
         Q_rxn_vol_av = Q_rxn_W / V
         Q_rev_vol_av = Q_rev_W / V
+        Q_cr_vol_av = Q_cr_W / V
```
Comment (valentinsulzer) [inline pybamm/models/submodels/thermal/base_thermal.py:232]: I think this term should be `I^2R/V` and we need to reverse engineer the previous terms to get to that. i.e. `Q_cr = I**2 / (n_elec * A) * R_contact`. Need input from @rtimms @brosaplanella @TomTranter about how to think about contact resistance heating in "volume-averaged" sense

**Review by valentinsulzer — CHANGES_REQUESTED** [review]

File: pybamm/models/submodels/thermal/base_thermal.py (hunk 1 above)
Comment (agriyakhetarpal) [reply pybamm/models/submodels/thermal/base_thermal.py]: another way is to do `pip install pre-commit` and then `pre-commit install` which will block changes that don't pass from being committed

Comment (parkec3) [reply pybamm/models/submodels/thermal/base_thermal.py]: Thanks. I updated the code to have the proper syntax.

File: pybamm/models/submodels/thermal/base_thermal.py (hunk 2 above)
Comment (rtimms) [reply pybamm/models/submodels/thermal/base_thermal.py:232]: I'm not sure the contact resistance heating makes sense outside of the lumped model - it's not obvious "where" the heating is taking place. For higher-dimensional models, it's certainly not the case that the contact resistance heating should appear as a uniform source everywhere in the cell.

**Review by valentinsulzer — CHANGES_REQUESTED** [review]
Code looks good but needs tests. The following tests would be good:
- a unit test in `base_lithium_ion_tests` that checks contact resistance + lumped thermal
- an integration test that runs a lumped thermal model with and without contact resistance heating and makes sure that the temperature is always larger when there is contact resistance heating

**Comment by rtimms** [comment]
It would also be good to update [this notebook](https://github.com/pybamm-team/PyBaMM/blob/develop/docs/source/examples/notebooks/models/thermal-models.ipynb) to include contact resistance heating in the lumped model. We should also update the docs of the thermal models to be more explicit about which heating terms are included.

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_thermal_models.py [hunk 3]
```diff
@@ -103,6 +103,56 @@ def err(a, b):
...
+        # get the average temperature from each model
+        avg_cell_temp = sols[0]["X-averaged cell temperature [K]"].entries
+        avg_cell_temp_cr = sols[1]["X-averaged cell temperature [K]"].entries
+
+        # check that the time-averaged cell temperature of the lumped thermal model
+        # with contact resistance is higher than without contact resistance
+        self.assertGreater(
+            sum(avg_cell_temp_cr) / len(avg_cell_temp_cr),
+            sum(avg_cell_temp) / len(avg_cell_temp),
+        )
```
Comment (valentinsulzer) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_thermal_models.py:154]: you can also use `np.testing.assert_array_less` to check the whole array

**Review by kratman — APPROVED** [review]

**Pull request merged**